network-rfc/11_portal_api.md

```diff
@@ -120,17 +136,22 @@ If a block `N` was included in the response, it is guaranteed that all the exist
 | The requested block range falls into the dataset range, but doesn't contain any blocks | `200` | Some chains may have gaps in the numbering. The request should not be retried in this case — no blocks will ever appear for such range. |
-| The first requested block is less than the beginning of the dataset range | `400` | The client should make sure the requested range is within the dataset range |
+| The first requested block is below the beginning of the dataset range | `400` | The client should make sure the requested range is within the dataset range |
 | The `parentBlockHash` in the request doesn't match the current parent hash for the given `fromBlock` | `409` | See below |
 
 In case of the `503` and `429` codes there may be an additional `Retry-After` header indicating the number of seconds to wait before retrying the request.
 
-### Real-time mode
-Before reaching the finalized head, the client may assume that there are no rollbacks (forks) and the blocks are ordered sequentially. `/finalized-stream` and `/stream` behaviour is the same for those ranges. After that, if the client wants to get the latest data, it should switch to the real-time mode and use the `/stream` endpoint. The above statements still apply for it, but there are additional considerations.
+### Handling reorgs (forks)
+
+Historical blockchain data is easy to deal with because one can assume that blocks are ordered sequentially and there is only one block for the given height (block number). However, when getting close to the blockchain head, the client has to consider the possibility of chain reorgs.
+
+As a hint about the finalization point, the portal exposes the `/finalized-head` endpoint. It responds with the last block which should be considered "final" meaning that it's unlikely to be reverted. For some datasets it's not possible to guarantee finality for sure, but if the client encounters a reorg deeper than the finalization point, it shouldn't take actions to resolve it automatically, and should signal the error to the user instead.
 
-With each query, the client should pass the `parentBlockHash` field of the known parent hash of the first requested block. The portal will check if the hash matches the parent hash of the current block with height `fromBlock` on the path to head. If it doesn't, the portal will respond with the `409` error with the list of the previous blocks on the path to the current head. In this case, the client should re-request the blocks starting from the last known ancestor. If no common ancestor was found (chains diverged too much), the client should repeat the procedure for the preceding blocks until the common ancestor is found.
+The portal additionally adds the `X-Sqd-Finalized-Head-Number` and `X-Sqd-Finalized-Head-Hash` headers to each successful response to save additional requests.
```

Comment:
It is not just about avoiding additional requests.

Assume `endpoint.finalized_head` was received from `/finalized-head`.

Then, when `client.head.number < endpoint.finalized_head.number` no finalization is possible, because there is no way to verify, that `endpoint.finalized_head` is a child of `client.head`.

In general, no finalization is possible when `endpoint.finalized_head` is not known to the client. Such situation is likely, because we skip blocks, that don't match the query.